order_book/src/order.rs

```diff
@@ -0,0 +1,289 @@
+                // is in the same token id as the base token, since it takes
+                // away from the amount of base tokens available for consumption
+                // by the fulfiller.
+                if num_required_outputs == 1 {
+                    if sci.required_output_amounts[0].token_id != base_token_id {
+                        return Err(Error::UnsupportedSci(format!(
+                        "Suspected required-change-output token id {} does not match partial output token id {}",
+                        sci.required_output_amounts[0].token_id, amount.token_id
+                    )));
+                    }
+                    max_base_amount = max_base_amount
+                        .checked_sub(sci.required_output_amounts[0].value)
+                        .ok_or_else(|| {
+                            Error::UnsupportedSci(format!(
+                                "max base amount {} is lower than required change {}",
+                                max_base_amount, sci.required_output_amounts[0].value
+                            ))
+                        })?;
+                }
+
+                (
+                    amount.token_id,
+                    min_base_amount..=max_base_amount,
+                    amount.value,
+                )
+            }
+            _ => {
+                return Err(Error::UnsupportedSci(format!(
+                    "Unsupported number of required/partial outputs {}/{}",
+                    input_rules.required_outputs.len(),
+                    input_rules.partial_fill_outputs.len()
+                )))
+            }
+        };
+
+        let id = OrderId::from(&sci);
+
+        let pair = Pair {
+            base_token_id,
+            counter_token_id,
+        };
+
+        Ok(Self {
+            sci,
+            id,
+            pair,
+            base_range,
+            max_counter_tokens,
+            timestamp,
+        })
+    }
+}
+
+impl Deref for Order {
+    type Target = SignedContingentInput;
+
+    fn deref(&self) -> &Self::Target {
+        &self.sci
+    }
+}
+
+impl Ord for Order {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // We sort orders by the following, in this order:
+        // 1) The pair (so that orders of the same pair are grouped together)
+        // 2) The ratio of base to counter, putting orders with a more favorable
+        // exchange rate (to the fulfiller) first.
+        // 3) Timestamp (so that older orders are filled first)
+        // 4) Order id (in case of orders where all the above were identical)
+
+        // The rate is calculated as base / counter. We want to sort by:
+        // (self_base / self_counter) > (other_base / other_counter)
+        // Since we want to avoid division, we multiply both sides by the denominators
+        // and get: self_base * other_counter > other_base * self_counter
+        let self_rate = other.max_base_tokens() as u64 * self.max_counter_tokens() as u64;
```

Comment:
```suggestion
        let self_rate = other.max_base_tokens() as u128 * self.max_counter_tokens() as u128;
```

I think you want this because we don't want to have wrap around if we overflow the u64 when we multiply